### PR TITLE
Tag groups can belong to groups

### PR DESCRIPTION
**Pull request by jbrw**
Allows tag groups to be assigned to groups other than the 'staff' group.

https://meta.discourse.org/t/allow-restriction-of-tags-to-non-staff-groups/161322

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/tag-groups-form.js [hunk 1]
```diff
@@ -4,32 +4,109 @@ import { isEmpty } from "@ember/utils";
...
+    );
+  },
+
+  @discourseComputed("buffered.permissions")
+  showPrivateChooser(permissions) {
+    if (!permissions) {
+      return true;
+    }
+
+    return permissions["everyone"] !== PermissionType.READONLY;
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/components/tag-groups-form.js:47]:
```suggestion
    return permissions.everyone !== PermissionType.READONLY;
```
It's preferred not to use the array syntax for javascript objects unless you have to.

File: app/assets/javascripts/discourse/app/components/tag-groups-form.js [hunk 2]
```diff
@@ -4,32 +4,109 @@ import { isEmpty } from "@ember/utils";
...
+
+  @discourseComputed("buffered.permissions", "allGroups")
+  selectedGroupIds(permissions, allGroups) {
+    if (!permissions || !allGroups) {
+      return [];
+    }
+
+    const selectedGroupNames = Object.keys(permissions);
+    let groupIds = [];
+    allGroups.forEach(function (group) {
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/components/tag-groups-form.js:58]:
```suggestion
    allGroups.forEach((group) => {
```
We prefer the fat arrow syntax unless you need a new `this` context.

File: app/assets/javascripts/discourse/app/components/tag-groups-form.js [hunk 3]
```diff
@@ -4,32 +4,109 @@ import { isEmpty } from "@ember/utils";
...
+
+    return groupIds;
+  },
+
+  everyoneSelected(permissions) {
+    if (!permissions) {
+      return true;
+    }
+
+    return permissions["everyone"] === PermissionType.FULL;
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/components/tag-groups-form.js:72]:
```suggestion
    return permissions.everyone === PermissionType.FULL;
```

File: app/assets/javascripts/discourse/app/components/tag-groups-form.js [hunk 4]
```diff
@@ -4,32 +4,109 @@ import { isEmpty } from "@ember/utils";
...
-        });
+        updatedPermissions.everyone = PermissionType.FULL;
       }
+
+      this.buffered.set("permissions", updatedPermissions);
+    },
+
+    setPermissionsGroups(groupIds) {
+      let permissions = {};
+      this.allGroups.forEach(function (group) {
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/components/tag-groups-form.js:103]:
```suggestion
      this.allGroups.forEach((group) => {
```

File: app/assets/javascripts/discourse/tests/acceptance/tag-groups-test.js [hunk 5]
```diff
@@ -36,3 +49,26 @@ test("tag groups can be saved and deleted", async (assert) => {
...
+
+    await visit("/tag_groups");
+    await click(".content-list .btn");
+
+    await fillIn(".tag-group-content h1 input", "test tag group");
+    await tags.expand();
+    await tags.selectRowByValue("monkey");
+
+    await click("#private-permission");
+    assert.ok(find(".tag-group-content .btn.btn-default")[0].disabled);
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/tests/acceptance/tag-groups-test.js:67]:
```suggestion
    assert.ok(find(".tag-group-content .btn.btn-default:disabled").length);
```
This is a huge nitpick but the query selector can find the disabled element which is slightly safer code than calling `[0]` which might not exist. I know it's a test, but still.

File: lib/discourse_tagging.rb [hunk 6]
```diff
@@ -296,12 +313,23 @@ def self.filter_allowed_tags(guardian, opts = {})
...
+      group_ids = permitted_group_ids(guardian)
+
+      builder.where(<<~SQL, group_ids, group_ids)
         id NOT IN (
-          SELECT tag_id
-            FROM tag_group_memberships tgm
-           WHERE tag_group_id NOT IN (SELECT tag_group_id FROM tag_group_permissions WHERE group_id = ?)
+          (SELECT tgm.tag_id
+             FROM tag_group_permissions tgp
+       INNER JOIN tag_groups tg ON tgp.tag_group_id = tg.id
```
Comment (eviltrout) [inline lib/discourse_tagging.rb:322]: The indentation here looks a bit off.

File: lib/discourse_tagging.rb [hunk 7]
```diff
@@ -348,16 +376,52 @@ def self.hidden_tag_names(guardian = nil)
...
         Group::AUTO_GROUPS[:everyone]
       )
+
+    query
+  end
+
+  def self.permitted_group_ids(guardian = nil)
+    group_ids = [Group::AUTO_GROUPS[:everyone]]
+
+    if guardian&.user.present?
```
Comment (eviltrout) [inline lib/discourse_tagging.rb:395]:
```suggestion
    if guardian.authenticated?
```

**Review by eviltrout — APPROVED** [review]
This looks great overall to me. I've suggested a bunch of minor changes. I assume you've tested it fairly comprehensively locally and everything is working well? If so feel free to merge when you're around for a bit in case it needs updating/changes.

File: lib/discourse_tagging.rb (hunk 6 above)
Comment (jbrw) [reply lib/discourse_tagging.rb:322]: The way the various chunks of hand-written SQL is formatted in this file is extremely inconsistent. Do we have a set of rules/style guide on how it should be done?

Comment (eviltrout) [reply lib/discourse_tagging.rb:322]: I agree it's inconsistent, but maybe this is displaying differently for you. Check this out:

image

The `INNER JOIN`s are to the left of the `(SELECT`, that they belong to. I feel like any parts of a subquery should be to the right of where it starts, or at least aligned in a column. The `FROM` and `WHERE` seem good.

**Pull request merged**